streaming/stream_transfer_task.cc

```diff
@@ -203,15 +203,27 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink) {
-                return repeat([&sink, si, got_error_from_peer] () mutable {
-                    return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
-                        if (mf && !(*got_error_from_peer)) {
+            bool partition_opened = false;
+            return do_with(std::move(sink), partition_opened, [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink, bool& partition_opened) {
+                return repeat([&sink, &partition_opened, si, got_error_from_peer] () mutable {
+                    return si->reader(db::no_timeout).then([&sink, &partition_opened, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
+                        if (*got_error_from_peer) {
+                            return make_exception_future<stop_iteration>(std::runtime_error("Got status error code from peer"));
+                        }
+                        if (mf) {
+                            if (mf->is_partition_start()) {
+                                partition_opened = true;
+                            } else if (mf->is_end_of_partition()) {
+                                partition_opened = false;
+                            }
```

Comment:
It would be simpler to replace `partition_opened` with `last_mutation_fragment_kind` and just do:
```C++
last_mutation_fragment_kind = mf->mutation_fragment_kind();
```
It would avoid two branches and thus would be less error-prone.

Reply:
I am thinking about if we should add more checks like below:

```
if (mf->is_partition_start()) {
   // Already opened?
   if (partition_opened) {
          throw
   }
    partition_opened = true;
} else if (mf->is_end_of_partition()) {
   // Already closed ?
    if (!partition_opened) {
          throw
    }
    partition_opened = false;
}
                                        

```

Reply:
This is starting to look like the `mutation_fragment_stream_validator` open-coded :).

Reply:
Is it possible to use mutation_fragment_stream_validator here?

Reply:
Yes. You would have to use the lower level validator, which is the one actually called `mutation_fragment_stream_validator` (the higher level one is called `mutation_fragment_stream_validating_filter`).
You can use it like this:
```c++
mutation_fragment_stream_validator validator(schema);
//...
// validates that the fragment kind montonicity, e.g. that after partition end the next fragment is partition start and not anything else.
if (!validator(mf->mutation_fragment_kind())) {
    throw;
}

// ... finally
if (!validator.on_end_of_stream()) {
    throw;
}
```
Note that the low level validator just returns `bool` from its methods and you have to do the error handling.

Reply:
This should be pretty cheap as it only validates the fragment kind validity, i.e. it will only compares enum values no keys are compared.

Reply:
Thanks. This looks good. If it is very cheap, I am thinking about if we should add the check inside  flat mutation reader.

Reply:
That would be quite redundant. We often build stack of readers half-a-dozen readers deep. This validation is cheap but not free, it would add up to be visible.

Reply:
Make sense.

Reply:
I am using the mutation_fragment_stream_validator in v2. Thanks!